cheroot/ssl/builtin.py

```diff
@@ -173,22 +239,164 @@ def get_environ(self, sock):
-                    ssl_environ.update(
-                        self.env_dn_dict(env_var, client_cert.get(cert_key)),
-                    )
+                ssl_environ['SSL_CLIENT_VERIFY'] = 'SUCCESS'
+                ssl_environ.update(
+                    self.env_cert_dict('SSL_CLIENT', client_cert),
+                )
+                ssl_environ['SSL_CLIENT_CERT'] = ssl.DER_cert_to_PEM_cert(
+                    sock.getpeercert(binary_form=True),
+                ).strip()
+
+        ssl_environ.update(self._server_env)
+
+        # not supplied by the Python standard library (as of 3.8)
+        # - SSL_SESSION_RESUMED
+        # - SSL_SECURE_RENEG
+        # - SSL_CLIENT_CERT_CHAIN_n
+        # - SRP_USER
+        # - SRP_USERINFO
 
         return ssl_environ
 
+    def loopback_for_cert(self, cert, private_key):
+        """Create a loopback connection to parse a cert with a private key."""
+        context = ssl.create_default_context(
+            cafile=self.certificate_chain,
+        )
+        context.load_cert_chain(cert, private_key)
+        context.check_hostname = False
+        context.verify_mode = ssl.CERT_NONE
+
+        # Python 3+ Unix, Python 3.5+ Windows
+        client, server = socket.socketpair()
+        try:
+            # ssl handshake is blocking and needs to happen on
+            # both ends at the same time so use a thread
+            thread = threading.Thread(
+                target=lambda: context.wrap_socket(
+                    server, do_handshake_on_connect=True,
+                    server_side=True,
+                ),
+            )
+            try:
+                thread.start()
+                return context.wrap_socket(
+                    client, do_handshake_on_connect=True,
+                    server_side=False,
+                ).getpeercert()
+            finally:
+                thread.join()
+        finally:
+            client.close()
+            server.close()
+
+    def parse_cert(self, cert, private_key):
+        """Parse a certificate."""
+        # loopback_for_cert uses socket.socketpair which was only
+        # introduced in Python 3.0 for *nix and 3.5 for Windows
+        # and requires OS support (AttributeError, OSError)
+        # it also requires a private key either in its own file
+        # or combined with the cert (SSLError)
+        with suppress(AttributeError, ssl.SSLError, OSError):
+            return self.loopback_for_cert(cert, private_key)
+
+        # KLUDGE: using an undocumented, private, test method to parse a cert
+        # unfortunately, it is the only built-in way without a connection
+        # as a private, undocumented method, it may change at any time
+        # so be tolerant of *any* possible errors it may raise
+        with suppress(Exception):
+            return ssl._ssl._test_decode_cert(cert)
+
+        return {}
+
+    def env_cert_dict(self, env_prefix, parsed_cert):
```

Comment:
Since this is a method, it's name should refer to an action rather than a state. Hence it should start with a verb.

Would this be an accurate suggestion?
```suggestion
    def make_env_cert_dict(self, env_prefix, parsed_cert):
```

Any alternatives? (`create_`/`convert_`/`generate_`?)
The goal is to have a semantically relevant identifier.

Reply:
Oh, and since it's only used from within the instance and isn't a part of the interface contract, it's a good idea to make it private:

```suggestion
    def _make_env_cert_dict(self, env_prefix, parsed_cert):
```